feat(quote): allow appending to an existing builder

tilpner opened the pull request:
This PR alters the biscuit-quote macros to support *appending* to existing builders, instead of creating new ones. The goal is to allow composition of facts and checks, instead of having to create the entire biscuit in one place.

It is currently possible to sort-of append biscuit builders, by using the `block!` macro and appending that to a `BiscuitBuilder`, but that causes avoidable allocations, and doesn't parse policies. With this PR,

```rust
b.append(block!(
  r#"
    fact(42);
  "#
));
```

becomes

```rust
biscuit!(
  &mut b,
  r#"
    fact(42);
  "#
);
```

- The macros evaluate to values of different types now, depending on if the caller passes a builder. With a builder it returns a `&mut B`, but a `B` without it.
- Distinguishing between clean creation and appending by looking at the first token type feels a bit wrong.
  - I didn't want to have duplicate versions of all three macros for appending.
  - Is there a better syntax?
- The implementation in `ParsedQuery` and `ParsedBiscuitQuery` is duplicated, and should probably be merged into one. I left it as-is for the moment, in case you want to evolve the syntax of them in different directions, but would be happy to adjust the PR to eliminate one of them.

divarvel commented:
The original macro design strove for consistency, at the cost of allocating a new builder when appending through a macro (regular fact, rules & checks can still be done). I'm not a fan of (unconstrained) function overloading in general, and macros tend to make things worse in this regard. otoh, unneeded allocations are bad.

I do see the merit of avoiding allocations though. Maybe adding macros for facts, rules & checks directly would allow to use both macros and still append directly without allocating a new block builder?

> It is currently possible to sort-of append biscuit builders, by using the block! macro and appending that to a BiscuitBuilder, but that causes avoidable allocations, and doesn't parse policies.

Policies can't be part of a block (in that case, the authority block), so the parsing part of the `biscuit!` macro is equivalent of the parsing part of the `block!` macro.

tilpner commented:
> I'm not a fan of (unconstrained) function overloading in general, and macros tend to make things worse in this regard. otoh, unneeded allocations are bad.

If you prefer separate macros (3 -> 6) and have a naming suggestion, I can implement those and try to share as much of the proc macro code as sensible.

> I do see the merit of avoiding allocations though. Maybe adding macros for facts, rules & checks directly would allow to use both macros and still append directly without allocating a new block builder?

I'm not sure what you mean. Could you show an example usage of those hypothetical macros?

> Policies can't be part of a block (in that case, the authority block), so the parsing part of the biscuit! macro is equivalent of the parsing part of the block! macro.

Yes, so since `Authorizer::append` takes a `BlockBuilder`, the pattern of `b.append(block!(...));` can't be used exclusively to iteratively construct authorizers. Unless I'm missing something, the policies either need to be part of the initial `authorizer!` invocation, or added with `Authorizer::add_policy`/`Authorizer::add_code` (both of which lag behind the macros in usability).

divarvel commented:
Right, i forgot that `Authorizer::append` was taking a `BlockBuilder`, I get what you mean :+1:

What I had in mind for the hypothetical macros was `block_builder.add_fact(fact!(r"fact(true)"));`, but that's not on par with what you've suggested. The current API for individual statements is returning `Result`s, so it's not convenient to use it for things that have guaranteed well-formedness.

All things considered, I think your proposed solution offers the best tradeoffs. The small input and return types inconsistencies can be covered with documentation and comprehensive examples. Separate macros for appending could work as well, but a clear naming would be paramount (since `append` is already used for appending blocks, it would be confusing to re-use it for something else). Maybe `add_to_biscuit!()` (or even `add_to_biscuit_authority!()`) / `add_to_block!()` / `add_to_authorizer!()` would work as well, idk.

As for the way the macro code is currently laid out, it was written in anger, so it could clearly be improved.

Geal commented:
I'm not a fan of having varying types on `biscuit!`. Could we consider a `BlockBuilder::merge` method instead? That way, multiple block parts can be created independently, some of them could even belong lived, then merged in one block that is appended to the token? I'm fine with `merge` either taking a `&mut self` or returning a new instance.

We should probably have an `AuthorizerBuilder` for consistency. Arguably, `AuthorizerPolicies` was supposed to be that, but right now it's not very usable

tilpner commented:
> Could we consider a BlockBuilder::merge method instead? 

Are you suggesting a `BlockBuilder::merge` function in addition to separate macros, or instead of them?

> That way, multiple block parts can be created independently, some of them could even belong lived, then merged in one block that is appended to the token?

I'm not so concerned with block reuse as I am with the unnecessary allocations. Consider for example

```rust
let mut b = biscuit!("...");

for fact in facts {
  // add some facts to the biscuit for every `fact`
  // ...
}
```

where facts might be queried from the database, and each loop iteration is supposed to add facts to the biscuit builder.
How would the loop body be implemented? Currently, it would have to `b.append(block!("..."));`, which will create three Vecs (some of them empty and thus cheap) per iteration, only to drain them into the corresponding Vecs of `b`, and then free the Vecs again.

Geal commented:
Would separate macros for fact, check etc make sense then? I was considering adding them for convenience

divarvel commented:
it would be nice, but currently there are no infallible methods on builders for already-built datalog elements, so they're a bit less nice than the existing block-level macros (the block-level macros call the `add_` methods and call unwrap themselves)

divarvel commented:
I think the best path forward is to:

- instead of overloading the existing `biscuit!()`, `block!()` and `authorizer!()` macros, introduce `biscuit_merge!()`, `block_merge!()`, and `authorizer_merge!()` macros
- rename `BlockBuilder::append` to `BlockBuilder::merge` since it's confusing wrt `Biscuit::append` (I can handle it in another PR, it's my bad naming choices, i can fix it myself)
- rename `Authorizer::append` to `Authorizer::merge`, make it take an `Authorizer` and have it return a Result (this way it can error if both authorizers are already carrying a token), and add a `Authorizer::merge_block` taking a `BlockBuilder`). Again, i can handle this part

I think this covers both ergonomic issues and allocation issues. wdyt @tilpner ?

tilpner commented:
> instead of overloading the existing biscuit!(), block!() and authorizer!() macros, introduce biscuit_merge!(), block_merge!(), and authorizer_merge!() macros

This is now implemented, and the macro implementation is a lot less duplicated. The `authorizer!` and `biscuit!` macros now additionally support the missing/extraneous parameter validation (previously only on `block!`), and parameter validation also accounts for parameters used in policies.

> rename BlockBuilder::append to BlockBuilder::merge since it's confusing wrt Biscuit::append

That's a good idea, I've renamed them.

> rename Authorizer::append to Authorizer::merge, make it take an Authorizer and have it return a Result (this way it can error if both authorizers are already carrying a token), and add a Authorizer::merge_block taking a BlockBuilder).

Done, but I'm not entirely sure I'm merging the correct fields. For now, in addition to merging the block builder, it also copies any policies, but there might be fields that should be merged, which currently aren't.

divarvel commented on biscuit-auth/src/token/authorizer.rs:
```diff
@@ -260,8 +260,13 @@ impl<'t> Authorizer<'t> {
-        self.authorizer_block_builder.append(other)
+    pub fn merge(&mut self, mut other: Authorizer) {
+        self.merge_block(other.authorizer_block_builder);
+        self.policies.append(&mut other.policies);
```
Thinking about it, a docblock stating that if `other` is already linked to a token, then it'll be discarded should be enough, no need to make `merge` return a `Result`

tilpner replied:
Right! I forgot to ask for clarification about that, as I didn't understand why that would be a problem :/

tilpner replied:
I've added a doc comment to point out that the token is discarded.

divarvel commented:
I really like how it looks like now

tilpner commented:
There are two things I'm still hoping to address:

- `set_macro_param` (and the subsequent `set_lenient`) are called for every item with every parameters. I don't think this is necessary, and we probably can provide just the relevant parameters
- @meh suggested supporting `let bar = 42; biscuit!("foo({bar}");` like the `println!`/`format!` macros do

tilpner commented:
Why is biscuit-quote using proc_macro instead of proc_macro2, and would you mind if I changed it to proc_macro2? It's the reason why `quote::__private::*` is needed.

divarvel commented:
The macro code was my first foray into rust macros, so it might not be up-to-date wrt best practices. Feel free to improve, I'll have a look afterwards to learn a couple things.

Thanks for the improvements!

tilpner commented:
I've updated both branches to use proc-macro2. This gets rid of the quote::__private shenanigans, and should be free as proc-macro2 was already an indirect dependency.

divarvel commented:
This looks really good! CI failures are spurious, it's because of the timeout check on token validation combined with poor perf on GHA workers.

tilpner commented:
I've relaxed the version requirement of proc-macro2 to just "1". I'm still unsure if you'd prefer [these changes](https://github.com/biscuit-auth/biscuit-rust/pull/109#issuecomment-1290543971) (https://github.com/tilpner/biscuit/commit/8426e4b6aac3993154243a9c914edfef1de45364) in a separate PR, or in this one (since it's related, but this one can also stand alone, if you think that makes for easier review).

Geal commented:
this looks great!

> There are two things I'm still hoping to address:
>
> - set_macro_param (and the subsequent set_lenient) are called for every item with every parameter. I don't think this is necessary, and we probably can provide just the relevant parameters

that can be improved but is not a big issue right now, there's not a huge amount of code generated

> - @meh suggested supporting let bar = 42; biscuit!("foo({bar}"); like the println!/format! macros do

that is a good idea, it can come in another PR

> - parameter value expressions are evaluated multiple times, leading to compile time errors if the parameter is used multiple times and the expression moves a value. this can be fixed by evaluating each parameter once, and then cloning as necessary
itmakes more sense to have the parameter move instead of hiding a clone, that would look more like what function calls taking ownership do

Geal reviewed (approved):
this is good to merge now, so we can either integrate it right now or wait for the remaining points

tilpner commented on biscuit-quote/src/lib.rs:
```diff
@@ -456,83 +456,184 @@ impl Builder {
-                .collect();
             Err(error::LanguageError::Parameters {
-                missing_parameters,
+                missing_parameters: Vec::new(),
```
There is no more fitting error variant for just unused parameters. For a better error message, one could be added.

tilpner commented on biscuit-quote/src/lib.rs:
```diff
@@ -456,83 +456,184 @@ impl Builder {
+                .map(|(name, _)| name.to_owned())
+                .collect(),
+            start: quote! {
+                let mut __biscuit_auth_item = #fact;
```
These prefixed names are necessary, so that in

```rust
biscuit!("foo({item}, {builder})", item = 42, builder = "foo");
```

the builder or constructed item aren't shadowing user parameters.

Geal replied:
right, that makes sense

Geal commented:
this looks good, once CI passes I'll merge it, thanks a lot!
